packages/core/src/utils/cursor.ts

```diff
@@ -0,0 +1,133 @@
+import type { AnyColumn, AnyColumnWithTable, OperandValueExpression, SelectQueryBuilder } from "kysely"
+import { z } from "zod"
+
+export interface PaginatedResult<T> {
+  data: T[]
+  next: Cursor | null
+}
+
+// Cursors are opaque strings that are parsed in the individual repositories
+export const CursorSchema = z.string()
+
+export const PaginateInputSchema = z
+  .object({
+    take: z.number(),
+    cursor: CursorSchema.optional().nullable(),
+  })
+  .optional()
+  .default({ take: 20, cursor: undefined })
+
+export type Pageable = z.infer<typeof PaginateInputSchema>
+export type Cursor = z.infer<typeof CursorSchema>
+
+type GetNextCursorOptions<T> = {
+  pageable: Pageable
+  buildCursor: (record: T) => Cursor
+}
+
+function getNextCursor<T>(records: T[], options: GetNextCursorOptions<T>): Cursor | null {
+  // We fetched take+1 records to determine if there is a record behind `take`.
+  // If we got exactly `take+1` records, there is a next page.
+  const hasNextPage = records.length === options.pageable.take + 1
+  if (!hasNextPage) {
+    return null
+  }
+
+  // biome-ignore lint/style/noNonNullAssertion: We know there is a record at -1 from the check above.
+  return options.buildCursor(records.at(-1)!)
```

Comment:
You can branch on `.at(take+1)` to avoid the noNonNullAssertions

Reply:
Good suggestion!